### DISCUSSION TITLE
Re-use resource route loader, as page component loader

### DISCUSSION THREAD

**Discussion by dan-cooke** [remix-run/remix#2241] [Q&A] [answered]
So i'm playing with the idea of hosting an API as a remix resource route. The benefits I can see to this approach are as follows:

1. Cut out the middleman, page renders will be fast because our API is running on the same machine as the frontend server.
2. Share type definitions directly without an intermediary types package
3. Flexibility to cut out the API fetch entirely and comm with the database in a page loader / action directly.

I would like to come up with a sensible pattern, with minimal boilerplate , duplication that allows me to maintain my api under `/routes/api/*` but ideally without having to perform a `fetch` from other `routes/` to hit the `/routes/api`

### Is it possible to re-use the `api/` loader / action, inside my page components.
It would be awesome If I can do something like this

**/routes/api/portfolios/index.tsx**
```/routes/api/portfolios/index.tsx
// My resource route to fetch some records from my DB
import { connectToAtlas } from '~/db/connect.server';
import { PortfolioModel } from '~/types';

// I like to be semantic with the names inisde the resource root
export const getPortfolios = async ({ request }) => {
  await connectToAtlas();

  return PortfolioModel.find();
};

export const loader = getPortfolios;

```

**/routes/portfolios**
```tsx
// import the same loader the api uses
import { getPortfolios } from '~/routes/api/portfolios`

export const loader = getPortfolios;

export default functoin PortfoliosPage() {

    const portfolios = useLoaderData();

    // page component

}
```
### But this breaks the client bundle, it seems to import some server code

The reason I would like to do this, is to reduce the boiler plate required to maintain an API route alongside your page loaders.

Currently I have to do this in order to hit my api loader 

```tsx
export const loader: LoaderFunction = async ({ request }) =>
  (await serverFetch(`/api/portfolios`)).json();
```

Which is more boiler plate, and also requires an extra piece of network IO (minor, but still)

I was wondering If anyone else has had much experience with this approach?

**Comment by kiliman** [comment, marked as answer]
I created a sample that reuses the API route from a normal route. 

https://codesandbox.io/s/reuse-api-route-k7i160?file=/app/routes/api/test.ts

One thing I did differently is that I don't export the named function, just the `loader`.
```ts
import { json, LoaderFunction } from 'remix'
const getData: LoaderFunction = async ({ request }) => {
  console.log('Server code is bundled')
  return json({ message: 'hello world' })
}
export const loader = getData
```
I then import the loader and reexport it from route.
```ts
import { useLoaderData } from 'remix'
import { loader } from '~/routes/api/test'
export { loader }

export default function Test() {
  const data = useLoaderData()
  return (
    <>
      <h1>Test</h1>
      <a href="/api/test">API</a>
      <pre>{JSON.stringify(data, null, 2)}</pre>
    </>
  )
}
```
I verified that the client bundle did not include any server code.
```ts
import "/build/_shared/chunk-5JTH7CC2.js";
import {
  React,
  init_react,
  useLoaderData
} from "/build/_shared/chunk-BAWAQCNH.js";

// browser-route-module:/sandbox/app/routes/test.tsx?browser
init_react();

// app/routes/test.tsx
init_react();
function Test() {
  const data = useLoaderData();
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement("h1", null, "Test"), /* @__PURE__ */ React.createElement("a", {
    href: "/api/test"
  }, "API"), /* @__PURE__ */ React.createElement("pre", null, JSON.stringify(data, null, 2)));
}
export {
  Test as default
};
```

**Reply by dan-cooke** [reply]
Wow what a subtle difference.

So it seems that `remix` will auto exclude any function named `loader` in the client bundle. ( I haven't confirmed this, just guessing from your example here)

This is a really nice pattern, with minimal boiler plate. 

I wonder if we can improve this further with an esbuild plugin

**Reply by dan-cooke** [reply]
I tried my hand at a quick esbuild plugin that would allow users to just write their `actions` and `loaders` in their pages.. and the plugin would create the relevant api routes.

But i've hit a roadblock with resolving imports - much trickier than I first imagined.

**Comment by sergiodxa** [comment]
Another way of doing that is to expose the `getPortfolios` function directly and not as loader and then calling it:

```ts
// app/services/get-portfolios.ts
import { connectToAtlas } from '~/db/connect.server';
import { PortfolioModel } from '~/types';

export const getPortfolios = async ({ request }) => {
  await connectToAtlas();

  return PortfolioModel.find();
};
```

```ts
// app/routes/api.portfolios.ts
import { getPortfolios } from "~/services/get-portfolios.ts"

export let loader: LoaderFunction = async ({ request }) => getPortfolios(request)
```

```tsx
// app/routes/portfolios.ts
import { getPortfolios } from "~/services/get-portfolios.ts"
import { PortfolioModel } from '~/types';

export let loader: LoaderFunction = async ({ request }) => getPortfolios(request)

export default function Screen() {
  let data = useLoaderData<PortfolioModel>()
  // render UI here
}
```

**Reply by dan-cooke** [reply]
Yeah! this is one approach I had used in the past, but I don't like the duplication of the `loader` signature.
